Supporting Coordinates Spaces in OSLObject

danieldresser opened the pull request:
This is some very early experimentation, but I feel like this will likely need some feedback from John about the approach, so I'm putting it up for initial comments sooner rather than later.

I need to do some object processing in Gaffer where I create prim vars based on the world space position.  I could probably hack this by manually passing an extra transform into some custom OSL, but it seems like it would be generally quite useful if the built-in ```transform( "world", P )``` worked.  I'm not sure if these changes could affect other users of this code, but I think it's still early enough in our use of this stuff at IE that it's probably easier to change now than later.

The first change is to call ```s->attribute( "commonspace", "object" );```.  This tells OSL that the default working space of OSLObject is object space, which makes sense.  Currently, it won't even try to convert to world space because it thinks everything is already in world space ( which it isn't ).

The next step is to implement get_matrix ( and probably get_inverse_matrix as well since we actually start with the object to world matrix, and it would be silly to inverse twice ).

This will require passing some extra data into renderstate ( or create a new objstate container, which doesn't appear necessary if we're creating a separate ShadingSystem per object anyway ).  It's not clear to me what the clean way to get this data passed in will be.  Any thoughts?

johnhaddon commented:
> I need to do some object processing in Gaffer where I create prim vars based on the world space position. I could probably hack this by manually passing an extra transform into some custom OSL, but it seems like it would be generally quite useful if the built-in transform( "world", P ) worked.

Hi Daniel, thanks for kicking this one off - it'll definitely be a good feature to have.

> The first change is to call s->attribute( "commonspace", "object" );. This tells OSL that the default working space of OSLObject is object space, which makes sense. Currently, it won't even try to convert to world space because it thinks everything is already in world space ( which it isn't ).

Quick question about this - does it affect the existing results of `getattribute()` at all? Are we telling OSL to put things in object space, or are we telling it we are giving it things in object space, and it should keep them there unless otherwise asked? Put another way, until now, should we actually have been converting to world space in our `RendererServices::get_userdata()` method?

> The next step is to implement get_matrix ( and probably get_inverse_matrix as well since we actually start with the object to world matrix, and it would be silly to inverse twice ).
> This will require passing some extra data into renderstate ( or create a new objstate container, which doesn't appear necessary if we're creating a separate ShadingSystem per object anyway ). It's not clear to me what the clean way to get this data passed in will be. Any thoughts?

I'm a bit confused about the "separate ShadingSystem per object" comment - do you mean separate RenderState per object? We should just be using a single static ShadingSystem, the one created in `shadingSystem()` in ShadingEngine.cpp.

Rather than think about the lowest level details first, I think it makes sense to start from the perspective of clients of the `GafferOSL::ShadingEngine` class. How should they provide the matrices to `ShadingEngine::shade()`?

The `shade()` method for a given `ShadingEngine` will be called from multiple threads, so we definitely don't want a mechanism involving stored state, so I think the simplest is to pass the extra information directly to `shade()`.

It seems like there are three obvious ways we could pass the information - either pass object-to-world, or pass a dictionary of transforms, or pass some sort of mini-renderer-services with a `matrix()` query. The only justification I can see for the complexity of the latter would be if computing the matrices was so prohibitively expensive that we'd want to do it on demand. I don't see any way of doing it on demand in OSLObject at present, because of the requirement to compute a hash before running the shader, and the lack of any `matrices_needed` attribute provided by OSL to let us know what the shader needs before querying it. Passing just object-to-world seems a bit limiting, because I could imagine us optionally wanting to pass the camera matrix (based on the current globals) and the matrices for CoordinateSystem objects too (or maybe even all objects matching a filter?). So some sort of dictionary makes the most sense to me. Something like this perhaps?

```
class ShadingEngine ...

    struct Transform
    {
        M44f fromObjectSpace;
        M44f toObjectSpace;
    }

    typedef map<string, Transform> Transforms;

    IECore::CompoundDataPtr shade( const IECore::CompoundData *points ) const;

```

With that passed in from the outside, it seems like it would be simple enough to store that in our per-object `RenderState` and then access it from our `RendererServices::get_matrix()`?  How does that sound? Let me know if I'm missing anything important or if I haven't explained myself well enough...

Cheers...
John

danieldresser commented:
> Quick question about this - does it affect the existing results of getattribute() at all?

As far as I know, not unless you call transform() explicitly, or explicitly declared a point in a specific space ( which wouldn't do anything currently anyway ).

> Are we telling OSL to put things in object space, or are we telling it we are giving it things in object space, and it should keep them there unless otherwise asked?

The latter.

> Put another way, until now, should we actually have been converting to world space in our RendererServices::get_userdata() method?

Yes.


> I'm a bit confused about the "separate ShadingSystem per object" comment

Oh, you're right, it was a bit late last night.  But I guess the key is that the shaderGlobals and renderState are per object, so we probably don't need to separate renderState and objState.

The API you're suggesting sounds quite reasonable.  I'm guessing based on your description that you intended for an additional parameter to shade(), though your code snippet just defines Transforms and doesn't use it anywhere?

I'll start taking a whack at this.

johnhaddon commented:
> The API you're suggesting sounds quite reasonable. I'm guessing based on your description that you intended for an additional parameter to shade(), though your code snippet just defines Transforms and doesn't use it anywhere?

Indeed - I'd intended to add a `const Transforms &transforms` argument.

> I'll start taking a whack at this.

Good stuff!

danieldresser commented:
OK, this now seems to do what I need.  See what you think.

johnhaddon commented on include/GafferOSL/ShadingEngine.h:
```diff
@@ -39,6 +39,7 @@
 #include "IECore/ObjectVector.h"
 
 #include "GafferOSL/TypeIds.h"
+#include "OpenImageIO/ustring.h"
```
We don't use OIIO or OSL types in our public APIs, so let's remove this.

johnhaddon commented on include/GafferOSL/ShadingEngine.h:
```diff
@@ -53,7 +54,15 @@ class ShadingEngine : public IECore::RefCounted
+			Imath::M44f toObjectSpace;
+		};
+
+		typedef std::map<OIIO::ustring, Transform> Transforms;
```
Once we've removed the OIIO header, we'll need to use either `string` or `IECore::InternedString` here.

johnhaddon commented on python/GafferOSLTest/OSLObjectTest.py:
```diff
@@ -100,6 +100,54 @@ def test( self ) :
+		p = GafferScene.Plane()
+		p["transform"]["translate"].setValue( IECore.V3f( 2, 0, 0 ) )
+
+
```
Unnecessary whitespace.

johnhaddon commented on src/GafferOSL/OSLObject.cpp:
```diff
@@ -147,8 +151,11 @@ void OSLObject::hashProcessedObject( const ScenePath &path, const Gaffer::Contex
+	h.append( inPlug()->fullTransformHash( path ) );
 }
 
+static const OIIO::ustring u_world("world");
```
We use `g_` for statics.

johnhaddon commented on src/GafferOSL/OSLObject.cpp:
```diff
@@ -183,7 +190,15 @@ IECore::ConstObjectPtr OSLObject::computeProcessedObject( const ScenePath &path,
+	ShadingEngine::Transforms transforms;
+
+	ShadingEngine::Transform world;
+	world.fromObjectSpace = inPlug()->fullTransform( path );
```
Do you think we should add a constructor to Transform? Maybe a two-matrix version and a one-matrix version that automatically computes the inverse?

johnhaddon commented on src/GafferOSL/ShadingEngine.cpp:
```diff
@@ -299,10 +299,33 @@ class RenderState
 	private :
 
 		size_t m_pointIndex;
+		const ShadingEngine::Transforms &m_transforms;
```
Since we're not going to use `ustring` in the public API, we'll need something else here. I'd suggest `boost::flat_map<ustring, Transform>`, since it seems to have pretty good lookup performance, especially for smallish numbers of keys (we use it in the Context).

johnhaddon replied:
The slower part of `flat_map` is insertion, but it does have a [fast constructor from presorted ranges]( http://www.boost.org/doc/libs/1_56_0/doc/html/boost/container/flat_map.html#idp29465184-bb), which is what we'll have from the `Transforms` coming via the constructor.

johnhaddon commented on src/GafferOSL/ShadingEngine.cpp:
```diff
@@ -299,10 +299,33 @@ class RenderState
 			m_pointIndex++;
 		}
+	
+		bool getMatrixToObject( OIIO::ustring name, Imath::M44f &result ) const
```
Generally we only prefix things with `get` if there's a corresponding `set`.

johnhaddon commented:
Except of course I forgot to check Travis, and all the ShadingEngine test cases are failing. Seems to be because of the new required transforms argument. Let's default that to a default constructed one on both the C++ and Python sides, so the tests pass without needing to be changed. We should also note the breaking change (added constructor argument) in the commit message so we can mention it in the release notes.

Thinking about it some more, I think it would have been nice to have two commits here - one making the changes to ShadingEngine, with accompanying tests in ShadingEngineTest, and then a second one doing the OSLObject work with the tests you have already.

danieldresser commented:
I've addressed most of your comments.  You suggested using flat_map because we could construct it from a presorted range - but the inputs would need to be sorted as OIIO::ustrings, which are ordered by memory address, so the data coming in can't be ordered properly?  I'm currently using boost::unordered_map, which is how OIIO handles dictionaries keyed by ustring.

For the public interface, I'm currently using a vector of name/transform pairs.  I can use a dictionary if you prefer, but because we never perform any lookups on this data before we have to build the ustring dictionary, it seems unnecessary.

I've made the transforms argument optional so that existing tests will pass - what I haven't done yet is exposed the new argument to Python properly.  What should I expect Python to pass in?  The easiest seems like it would be if I just took a dictionary of names to matrices, and automatically constructed the other matrix, since we can assume you don't care about setup performance if you're calling from Python.  Andrew suggested that in order to match conventions on C++ / Python parallelism I need to create a Python wrapper for the ShadingEngine::Transform class in Python, but he also said I should double check with you before doing that work.

danieldresser commented:
OK, I finally seem to have something working.  I've got no idea what I'm doing with this boost python stuff, so definitely take a close look at it.

The only thing I haven't addressed I seem to have broken the Mac build.  It failed with a link error which really doesn't seem to be my fault - Imath's matrix invert function depends on an exception that it can't find.  Do we need to explicitly link an extra Imath library on Mac?  Though wait, why is it happening now?  I'm not the only person calling M44f::inverse?

johnhaddon commented on src/GafferOSLModule/GafferOSLModule.cpp:
```diff
@@ -117,7 +118,71 @@ std::string oslCodeSource( const OSLCode &oslCode, const std::string &shaderName
+	);
+}
+
+IECore::CompoundDataPtr shadeWrapper( ShadingEngine &shadingEngine, const IECore::CompoundData *points, PyObject *obj_ptr = NULL )
```
I don't see anything _wrong_ with this, but I have a few suggestions for how it could be done a bit more easily, and also present a slightly better interface in Python :

- Use a `boost::python::dict pythonTransforms` argument instead of `PyObject *obj_ptr`. Then you don't need to worry about any of the `handle( borrowed() )` stuff, and don't have to cast to a dict. This also means the python help string will include the expected type.
- Use `boost::python::len( keys )` instead of `attr( "__len__" )`.

johnhaddon commented on src/GafferOSLModule/GafferOSLModule.cpp:
```diff
@@ -140,10 +205,21 @@ BOOST_PYTHON_MODULE( _GafferOSL )
+		;
+
+		class_<ShadingEngine::Transform>( "Transform" )
+			.def( "__init__", make_constructor( transformConstructor, default_call_policies() ) )
```
Is the `make_constructor` stuff actually needed here? Can you not just use `.def( init<const M44f &>() )` and `.def( init<const M44f &, const M44f&>() )`.

johnhaddon commented on src/GafferOSLModule/GafferOSLModule.cpp:
```diff
@@ -140,10 +205,21 @@ BOOST_PYTHON_MODULE( _GafferOSL )
+	{
+		scope s = IECorePython::RefCountedClass<ShadingEngine, IECore::RefCounted>( "ShadingEngine" )
+			.def( init<const IECore::ObjectVector *>() )
+			.def( "shade", &shadeWrapper, shadeOverloads() )
```
Rather than use `shadeOverloads()`, you can tell boost::python what the names and default values for the arguments are instead, and the python help string for the function will be more helpful. There's a decent example in the [Metadata bindings](https://github.com/GafferHQ/gaffer/blob/master/src/GafferBindings/MetadataBinding.cpp#L416).

danieldresser commented:
Hmm, I've addressed those comments ( for the record, some of the ugly Python dict binding stuff was copied from: https://github.com/ImageEngine/cortex/blob/master/src/IECorePython/CompoundDataBinding.cpp#L83, which I suppose should be updated as well ).

What still isn't working though is the Mac build - I've added the lib to the SConstruct - it looks right to me, but I'm still getting the same crash on travis machine.  Might be easier for you to debug with actual access to a mac?

johnhaddon commented:
> I've addressed those comments

Nice one - thanks.

> for the record, some of the ugly Python dict binding stuff was copied from: https://github.com/ImageEngine/cortex/blob/master/src/IECorePython/CompoundDataBinding.cpp#L83, which I suppose should be updated as well

Yeah, a lot has changed since those early Cortex days - boost improved and so did our understanding of it. Until we give the Cortex bindings a once over, I'd recommend Gaffer's bindings as a better reference.

> What still isn't working though is the Mac build - I've added the lib to the SConstruct - it looks right to me, but I'm still getting the same crash on travis machine. Might be easier for you to debug with actual access to a mac?

The error had moved from libGafferOSL.so to the _GafferOSL.so python module. I've added libIex to that, made a couple of minor formatting changes and merged. Thanks!
